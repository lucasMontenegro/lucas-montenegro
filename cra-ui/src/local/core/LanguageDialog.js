import React, { Fragment } from "react"
import { makeStyles } from "@material-ui/styles"
import Dialog from "@material-ui/core/Dialog"
import DialogTitle from "@material-ui/core/DialogTitle"
import DialogContent from "@material-ui/core/DialogContent"
import DialogActions from "@material-ui/core/DialogActions"
import Button from "@material-ui/core/Button"
import Typography from "@material-ui/core/Typography"
import List from "@material-ui/core/List"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"
import IconButton from "@material-ui/core/IconButton"
import Avatar from "@material-ui/core/Avatar"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { Link } from "local/core/links"
import makeTranslations from "local/makeTranslations"
const links = makeTranslations({
  en: `English`,
  es: `Español`,
}, `toArray`)
const labels = makeTranslations({
  en: `Language Menu`,
  es: `Menú de Idiomas`,
})
const titles = makeTranslations({
  en: `Select Language`,
  es: `Seleccione un Idioma`,
})
const closeText = makeTranslations({
  en: `CLOSE`,
  es: `CERRAR`,
})
const activeLinkLabels = makeTranslations({
  en: `Current language`,
  es: `Idioma actual`,
})
const useStyles = makeStyles(theme => ({
  dialogTitle: {
    display: `flex`,
    alignItems: `center`,
    "& > *": {
      margin: theme.spacing(1),
    },
  },
  avatar: {
    backgroundColor: theme.palette.primary.dark,
  },
}))
export default function LanguageDialog ({ state }) {
  const classes = useStyles()
  const icon = <FontAwesomeIcon icon={[`fas`, `language`]} />
  return (
    <Fragment>
      <IconButton
        aria-label={labels[state.languageCode]} id="open-language-dialog"
        aria-controls="language-dialog"         aria-haspopup="true"
        onClick={state.open}                    color="inherit"
      >
        {icon}
      </IconButton>
      <Dialog
        id="language-dialog"  aria-labelledby="language-dialog-title"
        onClose={state.close} open={state.isOpen}
      >
        <DialogTitle id="language-dialog-title" className={classes.dialogTitle} disableTypography>
          <Avatar id="language-dialog-avatar" className={classes.avatar}>{icon}</Avatar>
          <Typography component="h2" variant="h6">{titles[state.languageCode]}</Typography>
        </DialogTitle>
        <DialogContent>
          <nav>
            <List>
              {state.translations && links.map(({ languageCode, value }) => {
                const active = languageCode === state.languageCode
                return (
                  <ListItem
                    key={languageCode}
                    aria-label={active ? activeLinkLabels[languageCode] : undefined}
                    id={`language-link-${languageCode}`}
                  >
                    <ListItemText>
                      {active ? value : (
                        <Link onClick={state.close} to={state.translations[languageCode]}>
                          {value}
                        </Link>
                      )}
                    </ListItemText>
                  </ListItem>
                )
              })}
            </List>
          </nav>
        </DialogContent>
        <DialogActions>
          <Button id="close-language-dialog" onClick={state.close} color="primary">
            {closeText[state.languageCode]}
          </Button>
        </DialogActions>
      </Dialog>
    </Fragment>
  )
}