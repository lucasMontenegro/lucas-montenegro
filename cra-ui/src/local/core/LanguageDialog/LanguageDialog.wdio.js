import { expect } from "chai"
import supportedLanguages from "local/supportedLanguages"
describe(`local/core/LanguageDialog`, () => {
  it(`should render the button`, () => {
    supportedLanguages.forEach(languageCode => {
      browser.url(`/examples/core/LanguageDialog/${languageCode}`)
      const button = $(`#open-language-dialog`)
      expect(button.isDisplayed(), `#language-dialog-button isDisplayed`).to.be.true
      const msg = `#language-dialog-button getAttribute aria-label`
      expect(button.getAttribute(`aria-label`), msg).to.have.lengthOf.above(1)
    })
  })
  it(`should open`, () => {
    browser.url(`/examples/core/LanguageDialog/en`)
    const dialog = $(`#language-dialog`)
    expect(dialog.isDisplayed(), `#language-dialog isDisplayed`).to.be.false
    $(`#open-language-dialog`).click()
    dialog.waitForDisplayed(1000)
    expect(dialog.isDisplayed(), `#language-dialog isDisplayed`).to.be.true
  })
  it(`should render the dialog`, () => {
    supportedLanguages.forEach(languageCode => {
      browser.url(`/examples/core/LanguageDialog/${languageCode}`)
      $(`#open-language-dialog`).click()
      $(`#language-dialog`).waitForDisplayed(1000)
      function expectDisplayed (elem, selector) {
        elem.waitForDisplayed(1000)
        expect(elem.isDisplayed(), `${selector} isDisplayed`).to.be.true
      }
      function expectText (elem, selector) {
        expect(elem.getText(), `${selector} getText`).to.have.lengthOf.above(1)
      }
      function expectTextAttribute (elem, selector, attribute, expected) {
        const msg = `${selector} getAttribute ${attribute}`
        const { to } = expect(elem.getAttribute(attribute), msg)
        expected ? to.equal(expected) : to.have.lengthOf.above(1)
      }
      {
        const selector = `#language-dialog-title`
        const elem = $(selector)
        expectDisplayed(elem, selector)
        expectText(elem, selector)
      }
      {
        const selector = `#language-dialog-avatar`
        expectDisplayed($(selector), selector)
      }
      {
        const selector = `#close-language-dialog`
        const elem = $(selector)
        expectDisplayed(elem, selector)
        expectText(elem, selector)
      }
      const currentLanguage = languageCode
      supportedLanguages.forEach(languageCode => {
        const liSelector = `#language-link-${languageCode}`
        if (languageCode === currentLanguage) {
          const li = $(liSelector)
          expectDisplayed(li, liSelector)
          expectTextAttribute(li, liSelector, `aria-label`)
        } else {
          const selector = `${liSelector} > div > span > a`
          const anchor = $(selector)
          expectDisplayed(anchor, selector)
          const href = `/examples/core/LanguageDialog/${languageCode}`
          expectTextAttribute(anchor, selector, `href`, href)
        }
      })
    })
  })
  it(`should close from the modal`, () => {
    browser.url(`/examples/core/LanguageDialog/en`)
    const dialog = $(`#language-dialog`)
    $(`#open-language-dialog`).click()
    dialog.waitForDisplayed(1000)
    browser.keys(`Escape`)
    dialog.waitForDisplayed(1000, true)
    expect(dialog.isDisplayed(), `#language-dialog isDisplayed`).to.be.false
  })
  it(`should close from the close button`, () => {
    browser.url(`/examples/core/LanguageDialog/en`)
    const dialog = $(`#language-dialog`)
    $(`#open-language-dialog`).click()
    dialog.waitForDisplayed(1000)
    $(`#close-language-dialog`).click()
    dialog.waitForDisplayed(1000, true)
    expect(dialog.isDisplayed(), `#language-dialog isDisplayed`).to.be.false
  })
  it(`should close after clicking the link`, () => {
    supportedLanguages.forEach(languageCode => supportedLanguages.forEach(newLanguage => {
      if (newLanguage === languageCode) {
        return
      }
      browser.url(`/examples/core/LanguageDialog/${languageCode}`)
      $(`#open-language-dialog`).click()
      $(`#language-link-${newLanguage} > div > span > a`).click()
      const dialog = $(`#language-dialog`)
      dialog.waitForDisplayed(1000, true)
      expect(dialog.isDisplayed(), `#language-dialog isDisplayed`).to.be.false
    }))
  })
})