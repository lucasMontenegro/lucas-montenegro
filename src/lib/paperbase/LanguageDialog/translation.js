import Translation from "lib/Translation"
export default new Translation({
  en: {
    activeLinkLabel: `Current language`,
    openButtonLabel: `Language Menu`,
    title: `Select Language`,
    closeButtonText: `CLOSE`,
  },
  es: {
    activeLinkLabel: `Idioma actual`,
    openButtonLabel: `Men√∫ de Idiomas`,
    title: `Seleccione un Idioma`,
    closeButtonText: `CERRAR`,
  },
})