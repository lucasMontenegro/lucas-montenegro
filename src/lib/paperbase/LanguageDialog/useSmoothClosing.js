import { useRef } from "react"
// avoid language changes while closing the dialog
export default function useSmoothClosing (isOpen, children) {
  const wasOpen = useRef(false)
  const savedChildren = useRef(null)
  const result = (!isOpen && wasOpen.current) ? savedChildren.current : children
  wasOpen.current = isOpen
  savedChildren.current = children
  return result
}