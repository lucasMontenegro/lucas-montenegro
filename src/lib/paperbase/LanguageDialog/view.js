import React from "react"
import FontAwesomeIcon from "lib/fontAwesome/FontAwesomeIcon"
import IconButton from "@material-ui/core/IconButton"
import PropTypes from "prop-types"

import MuiDialog from "@material-ui/core/Dialog"
import { makeStyles } from "@material-ui/styles"

import DialogTitle from "@material-ui/core/DialogTitle"
import Avatar from "@material-ui/core/Avatar"
import Typography from "@material-ui/core/Typography"

import DialogContent from "@material-ui/core/DialogContent"
import List from "@material-ui/core/List"

import DialogActions from "@material-ui/core/DialogActions"
import Button from "@material-ui/core/Button"

import MuiListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"

import ReactRouterLink from "lib/links/ReactRouterLink"

const icon = <FontAwesomeIcon icon={[`fas`, `language`]} />
export function OpenButton ({ onClick, label }) {
  return (
    <IconButton
      id="open-language-dialog"
      aria-label={label}
      aria-controls="language-dialog"
      aria-haspopup="true"
      onClick={onClick}
      color="inherit"
    >
      {icon}
    </IconButton>
  )
}
OpenButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
}

export function Dialog ({ isOpen, onClose, children }) {
  return (
    <MuiDialog
      id="language-dialog"
      aria-labelledby="language-dialog-title"
      onClose={onClose}
      open={isOpen}
      scroll="body"
    >
      {children}
    </MuiDialog>
  )
}
Dialog.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}

const useTitleStyles = makeStyles(theme => ({
  dialogTitle: {
    display: `flex`,
    alignItems: `center`,
    "& > *": {
      margin: theme.spacing(1),
    },
  },
  avatar: {
    backgroundColor: theme.palette.primary.dark,
  },
}), { name: `language-dialog-title` })
export function Title ({ text, ...props }) {
  const classes = useTitleStyles()
  return (
    <DialogTitle
      {...props}
      id="language-dialog-title"
      className={classes.dialogTitle}
      disableTypography
    >
      <Avatar className={classes.avatar}>{icon}</Avatar>
      <Typography component="h2" variant="h6">{text}</Typography>
    </DialogTitle>
  )
}
Title.propTypes = {
  text: PropTypes.string.isRequired,
}

export function Content ({ links, ...props }) {
  return <DialogContent {...props}><nav><List>{links}</List></nav></DialogContent>
}
Content.propTypes = {
  links: PropTypes.array.isRequired,
}

export function Actions ({ onClose, closeButtonText, ...props }) {
  return (
    <DialogActions {...props}>
      <Button
        id="close-language-dialog"
        aria-controls="language-dialog"
        onClick={onClose}
        color="primary"
      >
        {closeButtonText}
      </Button>
    </DialogActions>
  )
}
Actions.propTypes = {
  onClose: PropTypes.func.isRequired,
  closeButtonText: PropTypes.string.isRequired,
}

export function ListItem ({ activeLabel, text }) {
  return (
    <MuiListItem>
      <ListItemText aria-label={activeLabel}>
        {text}
      </ListItemText>
    </MuiListItem>
  )
}
ListItem.propTypes = {
  activeLabel: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
}

export function ListItemLink ({ to, onClick, id, text }) {
  return (
    <MuiListItem>
      <ListItemText>
        <ReactRouterLink id={id} onClick={onClick} to={to}>
          {text}
        </ReactRouterLink>
      </ListItemText>
    </MuiListItem>
  )
}
ListItemLink.propTypes = {
  to: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
}