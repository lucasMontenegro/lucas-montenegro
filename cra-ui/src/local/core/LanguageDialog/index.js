import React, { Fragment } from "react"
import { withStyles } from "@material-ui/core/styles"
import Dialog from "@material-ui/core/Dialog"
import DialogTitle from "@material-ui/core/DialogTitle"
import DialogContent from "@material-ui/core/DialogContent"
import DialogActions from "@material-ui/core/DialogActions"
import Button from "@material-ui/core/Button"
import Typography from "@material-ui/core/Typography"
import List from "@material-ui/core/List"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"
import IconButton from "@material-ui/core/IconButton"
import Avatar from "@material-ui/core/Avatar"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { Link } from "local/core/links"
import translateLocation from "./translateLocation"
const links = [
  {
    languageCode: `en`,
    displayName: `English`,
  },
  {
    languageCode: `es`,
    displayName: `Español`,
  },
]
const labels = {
  en: `Language Menu`,
  es: `Menú de Idiomas`,
}
const titles = {
  en: `Select Language`,
  es: `Seleccione un Idioma`,
}
const closeText = {
  en: `CLOSE`,
  es: `CERRAR`,
}
const currentLanguageLabels = {
  en: `Current language`,
  es: `Idioma actual`,
}
export default withStyles(theme => ({
  dialogTitle: {
    display: `flex`,
    alignItems: `center`,
    "& > *": {
      margin: theme.spacing(1),
    },
  },
  avatar: {
    backgroundColor: theme.palette.primary.main,
  },
}))(class LanguageDialog extends React.Component {
  constructor (props) {
    super(props)
    this.state = { anchorEl: null }
    this.open = this.open.bind(this)
    this.close = this.close.bind(this)
  }
  open (event) {
    this.setState({ anchorEl: event.currentTarget })
  }
  close () {
    this.setState({ anchorEl: null })
  }
  render () {
    const { classes, languageCode, location, translations } = this.props
    const { anchorEl } = this.state
    const isOpen = Boolean(anchorEl)
    if (
        isOpen &&
        (!this.location || location !== this.location || languageCode !== this.languageCode)
      ) {
      this.location = location
      this.languageCode = languageCode
      this.locations = translateLocation(languageCode, location, translations)
    }
    const icon = <FontAwesomeIcon icon={[`fas`, `language`]} />
    return (
      <Fragment>
        <IconButton
          aria-label={labels[languageCode]}
          id="language-dialog-button"
          aria-controls="language-dialog"
          aria-haspopup="true"
          onClick={this.open}
          color="inherit"
        >
          {icon}
        </IconButton>
        <Dialog
          id="language-dialog"
          aria-labelledby="language-dialog-title"
          onClose={this.close}
          open={isOpen}
        >
          <DialogTitle
            id="language-dialog-title"
            disableTypography
            className={classes.dialogTitle}
          >
            <Avatar id="language-dialog-avatar" className={classes.avatar}>{icon}</Avatar>
            <Typography component="h2" variant="h6">{titles[languageCode]}</Typography>
          </DialogTitle>
          <DialogContent>
            <nav id="language-dialog-nav">
              <List>
                {this.locations && links.map(({ languageCode, displayName }) => {
                  const text = (
                    languageCode === this.languageCode ? (
                      <ListItemText aria-label={currentLanguageLabels[languageCode]}>
                        {displayName}
                      </ListItemText>
                    ) :
                    (
                      <ListItemText>
                        <Link to={this.locations[languageCode]}>{displayName}</Link>
                      </ListItemText>
                    )
                  )
                  return (
                    <ListItem key={languageCode}>{text}</ListItem>
                  )
                })}
              </List>
            </nav>
          </DialogContent>
          <DialogActions>
            <Button id="close-language-dialog" onClick={this.close} color="primary">
              {closeText[languageCode]}
            </Button>
          </DialogActions>
        </Dialog>
      </Fragment>
    )
  }
})