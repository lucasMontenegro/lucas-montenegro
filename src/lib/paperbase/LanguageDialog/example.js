import React from "react"
import { Route } from "react-router-dom"
import { ThemeProvider } from "@material-ui/styles"
import AppBar from "@material-ui/core/AppBar"
import Toolbar from "@material-ui/core/Toolbar"
import lightTheme from "lib/paperbase/themes/light"
import LanguageDialog from "lib/paperbase/LanguageDialog"
import languageDetector from "lib/languageDetector"
import CreateDestroy from "lib/utils/react/CreateDestroy"
const languageCodes = [`en`, `es`]
const rawLinks = [
  {
    languageCode: `en`,
    location: { pathname: `/paperbase/LanguageDialog/en` },
    otherProps: {
      key: `en`,
      id: `translate-to-en`,
      text: `English`,
    },
  },
  {
    languageCode: `es`,
    location: { pathname: `/paperbase/LanguageDialog/es` },
    otherProps: {
      key: `es`,
      id: `translate-to-es`,
      text: `Spanish`,
    },
  },
]
const links = languageCodes.reduce((links, currentLanguage) => {
  links[currentLanguage] = rawLinks.map(({ languageCode, location, otherProps }) => ({
    location: languageCode === currentLanguage ? null : location,
    otherProps,
  }))
  return links
}, {})
let count = 0
function getLinks () {
  console.log(`getLinks ${count}`)
  count++
  return links[languageDetector.get()]
}
function Example () {
  return (
    <ThemeProvider theme={lightTheme}>
      <AppBar><Toolbar><LanguageDialog getLinks={getLinks}/></Toolbar></AppBar>
    </ThemeProvider>
  )
}
function Init (props) {
  languageDetector.init(languageCodes)
  const ready = languageDetector.useReadyState()
  if (ready) {
    languageDetector.set(props.match.params.currentLanguage)
    return (
      <div style={{ paddingTop: 48 }}>
        <CreateDestroy Component={Example} />
      </div>
    )
  }
  return null
}
export default (<Route path="/paperbase/LanguageDialog/:currentLanguage" component={Init} />)