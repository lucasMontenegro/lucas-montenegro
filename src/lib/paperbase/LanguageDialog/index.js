import React, { useState, Fragment } from "react"
import translation from "lib/paperbase/LanguageDialog/translation"
import useSmoothClosing from "lib/paperbase/LanguageDialog/useSmoothClosing"
import {
  Title,
  Content,
  ListItemLink,
  ListItem,
  Actions,
  OpenButton,
  Dialog
} from "lib/paperbase/LanguageDialog/view"
import PropTypes from "prop-types"
export default function LanguageDialog ({ getLinks }) {
  const [isOpen, saveIsOpen] = useState(false)
  const close = () => saveIsOpen(false)
  const currentTranslation = translation.get()
  const children = useSmoothClosing(isOpen, [
    <Title key="title" text={currentTranslation.title} />,
    <Content
      key="content"
      links={getLinks().map(({ location, otherProps }) => {
        if (location) {
          return (
            <ListItemLink
              key={otherProps.key}
              to={location}
              onClick={close}
              id={otherProps.id}
              text={otherProps.text}
            />
          )
        }
        return (
          <ListItem
            key={otherProps.key}
            activeLabel={currentTranslation.activeLinkLabel}
            text={otherProps.text}
          />
        )
      })}
    />,
    <Actions key="actions" onClose={close} closeButtonText={currentTranslation.closeButtonText} />,
  ])
  return (
    <Fragment>
      <OpenButton onClick={() => saveIsOpen(true)} label={currentTranslation.openButtonLabel} />
      <Dialog isOpen={isOpen} onClose={close}>{children}</Dialog>
    </Fragment>
  )
}
LanguageDialog.propTypes = {
  getLinks: PropTypes.func.isRequired,
}