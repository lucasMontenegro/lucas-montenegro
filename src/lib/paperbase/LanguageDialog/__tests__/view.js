describe(`lib/paperbase/LanguageDialog/view`, () => {
  it(`should use the right dependency versions`, () => {
    expect(jestUtils.getDependencies([
      `react`,
      `@material-ui/core`,
      `prop-types`,
      `@material-ui/styles`,
    ])).toMatchSnapshot()
  })
  it(`should run`, () => {
    require("lib/paperbase/LanguageDialog/view")
  })
})