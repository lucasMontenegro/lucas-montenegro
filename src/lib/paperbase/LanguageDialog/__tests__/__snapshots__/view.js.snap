// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/paperbase/LanguageDialog/view should use the right dependency versions 1`] = `
Object {
  "@material-ui/core": "^4.5.0",
  "@material-ui/styles": "^4.5.0",
  "prop-types": "^15.7.2",
  "react": "^16.10.1",
}
`;
